fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateRewriter.java

```diff
@@ -0,0 +1,490 @@
+        for (ColumnRefOperator ref : allAggregateRefs) {
+            CallOperator call = aggregate.getAggregations().get(ref);
+            ColumnRefOperator newRef = factory.create(call.getFnName(), call.getType(), call.isNullable());
+            childContext.aggregations.put(newRef, call);
+
+            CallOperator newCall = genAggregation(call, newRef);
+            newAggregations.put(ref, newCall);
+        }
+
+        // group by
+        allRewrite.stream()
+                .map(a -> a.groupBys.keySet())
+                .flatMap(Collection::stream)
+                .filter(c -> aggregate.getGroupingKeys().contains(c))
+                .distinct().forEach(c -> childContext.groupBys.put(c, c));
+
+        LogicalAggregationOperator newAgg = LogicalAggregationOperator.builder().withOperator(aggregate)
+                .setAggregations(newAggregations).build();
+        optExpression = OptExpression.create(newAgg, optExpression.getInputs());
+        return processChild(optExpression, childContext);
+    }
+
+    private CallOperator genAggregation(CallOperator origin, ScalarOperator args) {
+        Function fn = Expr.getBuiltinFunction(origin.getFunction().getFunctionName().getFunction(),
+                new Type[] {args.getType()}, Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
+
+        Preconditions.checkState(fn instanceof AggregateFunction);
+        if (args.getType().isDecimalOfAnyVersion()) {
+            fn = DecimalV3FunctionAnalyzer.rectifyAggregationFunction((AggregateFunction) fn, args.getType(),
+                    origin.getType());
+        }
+
+        return new CallOperator(fn.getFunctionName().getFunction(), fn.getReturnType(),
+                Lists.newArrayList(args), fn);
+    }
+
+    @Override
+    public OptExpression visitLogicalJoin(OptExpression optExpression, AggregateContext context) {
+        if (isInvalid(optExpression, context)) {
+            return visit(optExpression, context);
+        }
+        LogicalJoinOperator join = (LogicalJoinOperator) optExpression.getOp();
+        if (join.getPredicate() != null) {
+            join.getPredicate().getUsedColumns().getStream().mapToObj(factory::getColumnRef)
+                    .forEach(v -> context.groupBys.put(v, v));
+        }
+
+        // push down aggregate
+        pushDownJoinAggregate(optExpression, context, 0);
+        pushDownJoinAggregate(optExpression, context, 1);
+        return optExpression;
+    }
+
+    private void pushDownJoinAggregate(OptExpression joinOpt, AggregateContext context, int child) {
+        LogicalJoinOperator join = (LogicalJoinOperator) joinOpt.getOp();
+        ColumnRefSet childOutput = joinOpt.inputAt(child).getOutputColumns();
+
+        AggregateContext childContext = new AggregateContext();
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : context.groupBys.entrySet()) {
+            if (childOutput.containsAll(entry.getValue().getUsedColumns())) {
```

Comment:
If this condition is not satisfied, the push down process may need to stop?

Reply:
No, the collector has decide which one can be pushed down, must be push down to left child or right child when go to here